ReplicaExchange/src/recore.jl

```diff
@@ -92,103 +99,194 @@ function re!(rng,
+                                                  i)
+          @inbounds states[GROUNDID] = samples[ub]
+        else
+          # If not ground state just return last sample
+          @inbounds states[i] = sim_chain_keep_last(rng, logenergys[i], states[i], samples_per_swap, i)
+        end
       end
     end
     swap_contexts!(rng, logenergys, states)
   end
   samples
 end
-# @pre length(logenergys) == length(states) "Must have one density per initial state"
 
-"Non-mutating re!"
+@pre re!(rng, logenergys, samples_per_swap, num_swaps, states, samples, sim_chain_keep_n) = num_swaps > 0 "There must be at least 1 swap"
+@pre re!(rng, logenergys, samples_per_swap, num_swaps, states, samples, sim_chain_keep_n) = all([isa(logenergy, Function) for logenergy in logenergys]) "logenergys is a Function"
+@pre re!(rng, logenergys, samples_per_swap, num_swaps, states, samples, sim_chain_keep_n) = isa(sim_chain_keep_n, Function) "sim_chain_keep_n is a Function"
+@pre re!(rng, logenergys, samples_per_swap, num_swaps, states, samples, sim_chain_keep_n) = length(logenergys) == length(states) "Must have one density per initial state"
+# usage of @cap and @ret in README of Spec.jl throws an error.
+# @post re!(rng, logenergys, samples_per_swap, num_swaps, states, samples, sim_chain_keep_n, sim_chain_keep_last, swap_contexts!) = (@cap(x), @ret) "Result is sorted version of input"
+
+
+"""
+  `re(rng,
+        logenergys,
+        samples_per_swap,
+        num_swaps,
+        states,
+        sim_chain_keep_n,
+        sim_chain_keep_last = last ∘ sim_chain_keep_n,
+        swap_contexts! = swap_contexts!)`
+             
+Replica Exchange Markov Chain Monte Carlo (REMCMC)
+Non-mutating version of [re!](@ref).
+
+Description:
+- REMCMC is a Markov Chain Monte Carlo algorithm for sampling from the cartesian product 
+  of a collection of `n` distributions given their respective densities.
+- REMCMC works by running `n` independent MCMC chains in parallel for `samples_per_swap` steps using 
+  a user-specified transition kernel (`sim_chain_keep_n`), and then subsequently proposing 
+  a special transition kernel that swaps the position of parallel chains.
+- REMCMC is agnostic to the user-specified transition kernel, as long as its stationary distribution
+  is equal to the distribution corresponding to its respective input density. E.g. this permits HMC or RW-MCMC.
+- A common use-case for REMCMC is when the target distribution is non-smooth and we would expect
+  a single MCMC transition kernel to become stuck near a local optima. Then, the collection of distributions
+  are induced by annealing the target distribution with a sequence of increasing temperatures.
+- In this implementation we assume that the user is only interested in samples from the single target distribution
+  and that the same `sim_chain_keep_n` is used for each of the `n` distributions.
+- Note: re does not mutate `states`.
```

Comment:
Don't think you need to repeat yourself (reference and maybe example are sufficient), see help on `sort!!` and `sort`